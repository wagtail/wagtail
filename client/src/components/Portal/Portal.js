import PropTypes from 'prop-types';
import { Component } from 'react';
import { createPortal } from 'react-dom';

/**
 * A Portal component which automatically closes itself
 * when certain events happen outside.
 * See https://reactjs.org/docs/portals.html.
 */
class Portal extends Component {
  constructor(props) {
    super(props);

    this.portal = document.createElement('div');

    this.onCloseEvent = this.onCloseEvent.bind(this);
  }

  onCloseEvent(e) {
    const { onClose } = this.props;

    if (!this.portal.contains(e.target)) {
      onClose();
    }
  }

  componentDidMount() {
    const { node, onClose, closeOnClick, closeOnType, closeOnResize } = this.props;

    node.appendChild(this.portal);

    if (closeOnClick) {
      document.addEventListener('mouseup', this.onCloseEvent);
    }

    if (closeOnType) {
      document.addEventListener('keyup', this.onCloseEvent);
    }

    if (closeOnResize) {
      window.addEventListener('resize', onClose);
    }
  }

  componentWillUnmount() {
    const { node, onClose } = this.props;

    node.removeChild(this.portal);

    document.removeEventListener('mouseup', this.onCloseEvent);
    document.removeEventListener('keyup', this.onCloseEvent);
    window.removeEventListener('resize', onClose);
  }

  render() {
    const { children } = this.props;

    return createPortal(children, this.portal);
  }
}

Portal.propTypes = {
  onClose: PropTypes.func.isRequired,
  node: PropTypes.instanceOf(Element),
  children: PropTypes.node,
  closeOnClick: PropTypes.bool,
  closeOnType: PropTypes.bool,
  closeOnResize: PropTypes.bool,
};

Portal.defaultProps = {
  node: document.body,
  children: null,
  closeOnClick: false,
  closeOnType: false,
  closeOnResize: false,
};

export default Portal;
