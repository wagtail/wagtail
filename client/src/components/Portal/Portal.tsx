/* eslint-disable react/static-property-placement */
import PropTypes from 'prop-types';
import { Component } from 'react';
import { createPortal } from 'react-dom';

/**
 * A Portal component which automatically closes itself
 * when certain events happen outside.
 * See https://reactjs.org/docs/portals.html.
 */
class Portal extends Component<{
  closeOnClick: boolean;
  closeOnResize: boolean;
  closeOnType: boolean;
  node: Element;
  onClose: () => void;
}> {
  static propTypes = {
    onClose: PropTypes.func.isRequired,
    node: PropTypes.instanceOf(Element),
    children: PropTypes.node,
    closeOnClick: PropTypes.bool,
    closeOnType: PropTypes.bool,
    closeOnResize: PropTypes.bool,
  };

  static defaultProps = {
    node: document.body,
    children: null,
    closeOnClick: false,
    closeOnType: false,
    closeOnResize: false,
  };

  portal: HTMLElement;

  constructor(props) {
    super(props);

    this.portal = document.createElement('div');

    this.onCloseEvent = this.onCloseEvent.bind(this);
  }

  onCloseEvent(event: MouseEvent) {
    const { onClose } = this.props;
    const target = event.target as Element;

    if (!this.portal.contains(target)) {
      onClose();
    }
  }

  componentDidMount() {
    const { node, onClose, closeOnClick, closeOnType, closeOnResize } =
      this.props;

    node.appendChild(this.portal);

    if (closeOnClick) {
      document.addEventListener('mouseup', this.onCloseEvent);
    }

    if (closeOnType) {
      document.addEventListener('keyup', this.onCloseEvent);
    }

    if (closeOnResize) {
      window.addEventListener('resize', onClose);
    }
  }

  componentWillUnmount() {
    const { node, onClose } = this.props;

    node.removeChild(this.portal);

    document.removeEventListener('mouseup', this.onCloseEvent);
    document.removeEventListener('keyup', this.onCloseEvent);
    window.removeEventListener('resize', onClose);
  }

  render() {
    const { children } = this.props;

    return createPortal(children, this.portal);
  }
}

export default Portal;
