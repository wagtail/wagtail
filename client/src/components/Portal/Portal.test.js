import React from 'react';
import { shallow, mount } from 'enzyme';
import Portal from './Portal';

const func = expect.any(Function);

describe('Portal', () => {
  beforeEach(() => {
    document.body.innerHTML = '';
  });

  it('empty', () => {
    expect(mount(<Portal onClose={() => {}} />)).toMatchSnapshot();
  });

  it('#children', () => {
    expect(mount(<Portal onClose={() => {}}>Test!</Portal>)).toMatchSnapshot();
  });

  it('component lifecycle', () => {
    document.removeEventListener = jest.fn();
    window.removeEventListener = jest.fn();

    const wrapper = mount(<Portal onClose={() => {}}>Test!</Portal>);

    expect(document.body.innerHTML).toMatchSnapshot();

    wrapper.instance().componentWillUnmount();

    expect(document.body.innerHTML).toBe('');

    expect(document.removeEventListener).toHaveBeenCalledWith('mouseup', func);
    expect(document.removeEventListener).toHaveBeenCalledWith('keyup', func);
    expect(window.removeEventListener).toHaveBeenCalledWith('resize', func);

    document.removeEventListener.mockRestore();
    window.removeEventListener.mockRestore();
  });

  describe('#onClose', () => {
    beforeEach(() => {
      jest.spyOn(document, 'addEventListener');
      jest.spyOn(window, 'addEventListener');
    });

    afterEach(() => {
      document.removeEventListener.mockRestore();
      window.removeEventListener.mockRestore();
    });

    it('#closeOnClick', () => {
      const onClose = jest.fn();
      shallow(
        <Portal onClose={onClose} closeOnClick>
          Test!
        </Portal>
      );
      expect(document.addEventListener).toHaveBeenCalledWith('mouseup', func);
    });

    it('#closeOnType', () => {
      const onClose = jest.fn();
      shallow(
        <Portal onClose={onClose} closeOnType>
          Test!
        </Portal>
      );
      expect(document.addEventListener).toHaveBeenCalledWith('keyup', func);
    });

    it('#closeOnResize', () => {
      const onClose = jest.fn();
      shallow(
        <Portal onClose={onClose} closeOnResize>
          Test!
        </Portal>
      );
      expect(window.addEventListener).toHaveBeenCalledWith('resize', func);
    });
  });

  describe('onCloseEvent', () => {
    it('should close', () => {
      const onClose = jest.fn();
      const wrapper = mount(<Portal onClose={onClose}>Test!</Portal>);
      const target = document.createElement('div');

      wrapper.instance().onCloseEvent({ target });

      expect(onClose).toHaveBeenCalled();
    });

    it('not should close', () => {
      const onClose = jest.fn();
      const wrapper = mount(
        <Portal onClose={onClose}>
          <div id="test">Test</div>
        </Portal>
      );
      const target = document.querySelector('#test');

      wrapper.instance().onCloseEvent({ target });

      expect(onClose).not.toHaveBeenCalled();
    });
  });
});
