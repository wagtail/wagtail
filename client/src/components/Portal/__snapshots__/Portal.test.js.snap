// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal #children 1`] = `
<Portal
  closeOnClick={false}
  closeOnResize={false}
  closeOnType={false}
  node={
    <body>
      <div>
        Test!
      </div>
    </body>
  }
  onClose={[Function]}
>
  Test!
</Portal>
`;

exports[`Portal component lifecycle 1`] = `"<div>Test!</div>"`;

exports[`Portal empty 1`] = `
<Portal
  closeOnClick={false}
  closeOnResize={false}
  closeOnType={false}
  node={
    <body>
      <div />
    </body>
  }
  onClose={[Function]}
/>
`;
